### PR TITLE
Add explicit domain rule

### PR DESCRIPTION
**Pull request by JakaMohorko**
Adds a new Structure type called `ExplicitDomainRuleStructure` with a child `MinExpectedDelta`. The structure is used to provide clients with information about the expected rate of asynchronous signals. This allows them to predict the required buffer sizes.

The expected delta specifies the expected "raw" distance between two samples. Eg. if the signal has domain data in integer ticks, the expected delta should be in said ticks as well. 

If this field is not provided by a data descriptor, I see 2 possibilities:
- The "BaseRuleDescriptionStructure" should be used, as it was up until now for explicit signals
- The field should be set to -1

### CODE REVIEW

**Comment by ml-hbk** [comment]
It will be used for domain signal only but does it need to be in the type name? Why not calling it ExplicitRuleStructure? Looking at the existing types ExplicitRuleDescriptionStructure.

I would prefer to use BaseRuleDescriptionStructure when there is no such MinExpectedDelta.

**Comment by JakaMohorko** [comment]
> It will be used for domain signal only but does it need to be in the type name? Why not calling it ExplicitRuleStructure? Looking at the existing types ExplicitRuleDescriptionStructure.
> 
> I would prefer to use BaseRuleDescriptionStructure when there is no such MinExpectedDelta.

We currently do not have a type called "ExplicitRuleDescriptionStructure" unless I missed it.  From what I can tell we use the "BaseRuleDescriptionStructure" with the type name "Explicit".  The reason I included "Domain" in the name, is mostly that this delta serves no real purpose for signals that do not define domain data. The field is used basically to determine the expected sample rate. As such I hesitate to add a non-optional (OPC UA structs should really avoid having optionals...) field to an ExplicitRuleStructure that all signals would then use.

**Comment by ml-hbk** [comment]
> We currently do not have a type called "ExplicitRuleDescriptionStructure" unless I missed it. From what I can tell we use the "BaseRuleDescriptionStructure" with the type name "Explicit". The reason I included "Domain" in the name, is mostly that this delta serves no real purpose for signals that do not define domain data. The field is used basically to determine the expected sample rate. As such I hesitate to add a non-optional (OPC UA structs should really avoid having optionals...) field to an ExplicitRuleStructure that all signals would then use.

ok, understood.

**Comment by ml-hbk** [comment]
I just had a talk with a colleague. In QuantumX we do have kind of an optional timeout for such signals. Would be MaxExpectedDelta. Client could notify timeout if this rule is broken. If we have both, there would need to be the possibility to tell that the parameter is to be ignored. Instead of -1 I would opt for 0.

**Comment by JakaMohorko** [comment]
> > We currently do not have a type called "ExplicitRuleDescriptionStructure" unless I missed it. From what I can tell we use the "BaseRuleDescriptionStructure" with the type name "Explicit". The reason I included "Domain" in the name, is mostly that this delta serves no real purpose for signals that do not define domain data. The field is used basically to determine the expected sample rate. As such I hesitate to add a non-optional (OPC UA structs should really avoid having optionals...) field to an ExplicitRuleStructure that all signals would then use.
> 
> ok, understood.

Also note that this is supplemental information that signals could provide, and is not required. It just makes the job easier for clients to store the signal data. It could be added within a "parameters" field, but we don't have such a field for base rules. Ergo why I added a new structure type in order to not break any current implementations.

The SDK will not have this field as a required field - it'll be optional.

**Comment by JakaMohorko** [comment]
> I just had a talk with a colleague. In QuantumX we do have kind of an optional timeout for such signals. Would be MaxExpectedDelta. Client could notify timeout if this rule is broken. If we have both, there would need to be the possibility to tell that the parameter is to be ignored. Instead of -1 I would opt for 0.

Honestly, this would be up to the client to interpret or ignore. I would like these fields to be fully optional, and only be supplementary info for clients to use. For the MaxExpectedDelta, we were thinking of per-signal keep-alive pings with Dejan, but haven't really expended on the idea. The MinExpectedDelta comes from Dewesoft where we need it  to estimate the circular buffer size requirements for signals.

And both 0 and -1 work. If we add 2 parameters, we need to have such a value to indicate that a field should be ignored.

**Review by nilsRoettger — APPROVED** [review]
It was discussed in the open discussion. The expectedMinDelta is fine for me.

File: opendaq/Opc.Ua.Daq.Bsp.NodeSet2.xml [hunk 1]
```diff
@@ -181,8 +181,8 @@
         <DisplayName>DaqAxisInformation</DisplayName>
         <References>
             <Reference ReferenceType="HasTypeDefinition">i=69</Reference>
-            <Reference ReferenceType="HasDescription" IsForward="false">ns=1;i=5024</Reference>
```
Comment (nilsRoettger) [inline opendaq/Opc.Ua.Daq.Bsp.NodeSet2.xml:184]: One question: Could you modify the file in that way, that these things are not shown as changes.

**Pull request merged**